### PR TITLE
Recommission api level support

### PR DESCRIPTION
**Pull request by pranikum**
### Description
This PR is taken from https://github.com/opensearch-project/OpenSearch/pull/4321. 
Some checks were failing in above PR. Required to fix the previous commits. Which was creating issue. Hence created a separate PR.

This change contain REST API layer changes for Recommission API support.
This pr is dependent of Service layer changes. (https://github.com/opensearch-project/OpenSearch/pull/4320)

This change has some changes for the PR https://github.com/opensearch-project/OpenSearch/pull/4084

### Issues Resolved
[[List any issues this PR will resolve]](https://github.com/opensearch-project/OpenSearch/issues/4083)

### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [X] Commits are signed per the DCO using --signoff
- [X] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by imRishN** [comment]
LGTM. Please merge this only after service level changes for recommission is merged

**Review by imRishN — APPROVED** [review]

File: rest-api-spec/src/main/resources/rest-api-spec/api/cluster.delete_decommission_awareness.json
Comment (saratvemulapalli) [inline rest-api-spec/src/main/resources/rest-api-spec/api/cluster.delete_decommission_awareness.json:4]: This link redirects to docs. Is the documentation work in progress? or Am I missing something?

File: server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/delete/DeleteDecommissionStateRequest.java
Comment (saratvemulapalli) [inline server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/delete/DeleteDecommissionStateRequest.java:23]: This class literally has no data. Is there a better way to do this? 
May be just define a generic request?

File: server/src/test/java/org/opensearch/rest/action/admin/cluster/RestDeleteDecommissionStateActionTests.java
Comment (saratvemulapalli) [inline server/src/test/java/org/opensearch/rest/action/admin/cluster/RestDeleteDecommissionStateActionTests.java:42]: Where is this being used?

**Review by saratvemulapalli — COMMENTED** [review]
Looks like a lot of boiler plate code, which is absolutely ok but could you help understand:
- how this will be hooked up?
- Looks like the API/feature is experimental, do you plan on to gate it by a feature flag?

File: server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/delete/DeleteDecommissionStateRequest.java
Comment (pranikum) [reply server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/delete/DeleteDecommissionStateRequest.java:23]: yes. you are right. But we will be enhancing this class further. Hence we will need to have a separate object.

File: server/src/test/java/org/opensearch/rest/action/admin/cluster/RestDeleteDecommissionStateActionTests.java
Comment (pranikum) [reply server/src/test/java/org/opensearch/rest/action/admin/cluster/RestDeleteDecommissionStateActionTests.java]: Removed. Thanks for pointing out.

File: server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/delete/DeleteDecommissionStateResponse.java
Comment (anshu1106) [inline server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/delete/DeleteDecommissionStateResponse.java:23]: Why do we need to implement ```ToXContentObject```?

File: server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/delete/TransportDeleteDecommissionStateAction.java
Comment (anshu1106) [inline server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/delete/TransportDeleteDecommissionStateAction.java:80]: is it better to put this in debug log?

File: rest-api-spec/src/main/resources/rest-api-spec/api/cluster.delete_decommission_awareness.json
Comment (anshu1106) [inline rest-api-spec/src/main/resources/rest-api-spec/api/cluster.delete_decommission_awareness.json:2]: lets have consistent naming across.

Comment (pranikum) [reply rest-api-spec/src/main/resources/rest-api-spec/api/cluster.delete_decommission_awareness.json:2]: keeping it similar to ```get_decommission_awareness"```

File: server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/delete/DeleteDecommissionStateResponse.java
Comment (pranikum) [reply server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/delete/DeleteDecommissionStateResponse.java]: Removed. Thanks for pointing.

File: server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/delete/TransportDeleteDecommissionStateAction.java
Comment (pranikum) [reply server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/delete/TransportDeleteDecommissionStateAction.java:80]: Keeping it as info only. Don't expect call to decommission/recommission be called frequently. It will be a nice checkpoint to track.

**Comment by anshu1106** [comment]
API changes look good to me. Will approve once service changes are merged and api changes are updated.

File: server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/delete/TransportDeleteDecommissionStateAction.java
Comment (Bukhtawar) [reply server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/delete/TransportDeleteDecommissionStateAction.java:80]: please make it more verbose with the request details. This PR is still WIP, please also rename the PR appropriately

File: server/src/main/java/org/opensearch/rest/action/admin/cluster/RestDeleteDecommissionStateAction.java
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/rest/action/admin/cluster/RestDeleteDecommissionStateAction.java:47]: If there are multiple attributes we should just clear only that attribute. Hope that has been factored in

Comment (pranikum) [reply server/src/main/java/org/opensearch/rest/action/admin/cluster/RestDeleteDecommissionStateAction.java:47]: Currently we are deleting the decommissionAttribute of any type which is set. Also as per now only one decommission attribute can be set. So we are clearing that one only.
Creating a github issue to handle specific attribute. Will add the Github issue here.

Comment (pranikum) [reply server/src/main/java/org/opensearch/rest/action/admin/cluster/RestDeleteDecommissionStateAction.java:47]: Create an issue
https://github.com/opensearch-project/OpenSearch/issues/4654

File: server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/delete/TransportDeleteDecommissionStateAction.java
Comment (pranikum) [reply server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/delete/TransportDeleteDecommissionStateAction.java]: Added request object in the log.

File: rest-api-spec/src/main/resources/rest-api-spec/api/cluster.delete_decommission_awareness.json
Comment (Bukhtawar) [inline rest-api-spec/src/main/resources/rest-api-spec/api/cluster.delete_decommission_awareness.json:7]: why is it experimental?

**Review by Bukhtawar — COMMENTED** [review]
Can we add integ tests as a part of this change for this API

File: rest-api-spec/src/main/resources/rest-api-spec/api/cluster.delete_decommission_awareness.json
Comment (pranikum) [reply rest-api-spec/src/main/resources/rest-api-spec/api/cluster.delete_decommission_awareness.json:7]: We will be marking this to stable once all meta issues are resolved we already have task identified for same.
https://github.com/opensearch-project/OpenSearch/issues/3957

**Comment by pranikum** [comment]
> Can we add integ tests as a part of this change for this API



> Can we add integ tests as a part of this change for this API

[pranikum] : Planning to add integ test as a separate pr once this is merged. Already have done that.. But we have few more changes to go around this flow. Let me know if this looks fine. This will help in avoiding update of integration tests if that gets affected due to other changes around this are if any.

**Review by Bukhtawar — APPROVED** [review]
We might need integ tests for this

**Comment by pranikum** [comment]
> We might need integ tests for this

Adding the link to previous comment
https://github.com/opensearch-project/OpenSearch/pull/4604#issuecomment-1271245484

**Pull request merged**
